### ISSUE TITLE
BUG: online doc returns 404 - wrong `reference` in url

### ISSUE THREAD

**Issue by prisae** [scipy/scipy#14267]
The new theme seems to have introduced a duplicate `reference` in the URL, e.g.,

- Before: https://docs.scipy.org/doc/scipy/reference/linalg.interpolative.html
- After: https://docs.scipy.org/doc/scipy/reference/reference/linalg.interpolative.html

The effect was that today all SciPy search engine results I found were broken.

The search engines will update eventually. But the double `/reference/reference/` in the url is unnecessary, so it might be better removed?

**Comment by adeak** [comment]
This was also noted on the mailing list. @rgommers [said](https://mail.python.org/pipermail/numpy-discussion/2021-June/081898.html) he'd try to tackle it.

On the same topic, small issue I noticed: Release Notes still claims 1.7.0 is not released yet (note at the top) https://scipy.github.io/devdocs/release.1.7.0.html. Didn't seem worth opening a new issue for.

**Comment by prisae** [comment]
Must have missed the message of the mailing list then, sorry.

I'll leave it open for your point as well @adeak, but feel free to close it @rgommers

**Comment by ilayn** [comment]
The announcement is merged https://github.com/scipy/scipy.org/pull/396 but apparently the site didn't get refreshed. It will show up soon. So I'll close this since Ralf is probably going to see this 

Thanks all for your considerations.

**Comment by adeak** [comment]
@prisae seems that the corresponding messages were sent to scipy-user and numpy-discussion (I only saw the latter).

**Comment by prisae** [comment]
Thanks, yes I double checked it as well. Seems I am not on either of these two...

**Comment by ilayn** [comment]
@tupui is there any place where we can have a look at the doubling problem closely as it is picking up some traction? Maybe past related PRs where it might have been introduced

**Comment by rgommers** [comment]
Pretty sure it's an uploading issue; the devdocs don't have the duplicate `reference/reference`.

**Comment by rgommers** [comment]
It's not an uploader bug, the structure actually changed: 

https://docs.scipy.org/doc/scipy-1.6.3/reference/
https://docs.scipy.org/doc/scipy-1.6.3/reference/tutorial/index.html
https://docs.scipy.org/doc/scipy-1.6.3/reference/cluster.html

https://docs.scipy.org/doc/scipy-1.7.0/reference/
https://docs.scipy.org/doc/scipy-1.7.0/reference/tutorial/index.html#user-guide
https://docs.scipy.org/doc/scipy-1.7.0/reference/reference/cluster.html (note the double reference)

Fuzzy on why we didn't notice this earlier. We also can't drop the `reference/` prefer - that will fix the API doc links, but break the tutorial ones.

Reopening.

**Comment by rgommers** [comment]
API docs are more important than tutorial though, so I'm tempted to drop the `reference/` anyway, it doesn't quite make sense. Copying things over on the server to `..` will give us double copies of everything, but seems like a good band aid to fix all links. Any concerns with doing that?

**Comment by rgommers** [comment]
The offending line is the last one here (from `doc/Makefile` in this repo):
```
doc-dist: html-scipyorg html
	test -d build/latex || make latex
	make -C build/latex all-pdf LATEXOPTS="-halt-on-error"
	-test -d build/htmlhelp || make htmlhelp-build
	-rm -rf build/dist
	mkdir -p build/dist
	cp -r build/html-scipyorg build/dist/reference
```

**Comment by stefanv** [comment]
The tutorial url makes more sense without the `reference` in it anyway, so +1 on simply copying to `build/dist`.

**Comment by rgommers** [comment]
Okay, `cp -R . ..` run on the server. That fixes the broken links to the API docs. I am changing the `Makefile` and rebuilding, to push up a clean doc build and ensure we have the proper fix in place.

**Comment by adeak** [comment]
Should the rebuild also fix the "Scipy 1.7.0 is not released yet!" note in the release notes, or should I open a new issue to track that?

**Comment by rgommers** [comment]
I'll fix that too, thanks for the suggestion.

**Comment by tylerjereddy** [comment]
I'm not so sure the current fix does what we want either, since it broke the docs a little I think.

![image](https://user-images.githubusercontent.com/7903078/127803254-d30f47d1-eb87-4cff-8ef6-c2ade2593dbf.png)

![image](https://user-images.githubusercontent.com/7903078/127803273-61569cf7-a67b-4dba-bd58-aba631c45d56.png)

I'll try to do a bit of manual cleaning for now, but not sure we want to forward port that precisely just yet

**Comment by tylerjereddy** [comment]
Here is what I did to "fix" manually on top of the cross-referenced PR:

`cd /srv/docs_scipy_org/doc/scipy-1.7.1`
`cp -a html-scipyorg/. .`

**Comment by melissawm** [comment]
I am getting a 404 on building the NumPy docs because of the SciPy objects.inv file:

```
WARNING: failed to reach any of the inventories with the following issues:
intersphinx inventory 'https://docs.scipy.org/doc/scipy/reference/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 404 Client Error: Not Found for url: https://docs.scipy.org/doc/scipy/reference/objects.inv
```

I am guessing this is related to this issue but not sure...

**Comment by ilayn** [comment]
Probab;y once the url problem is fixed definitely the object files have to be recompiled.

**Comment by rgommers** [comment]
It doesn't look fixed to me after the manual changes; neither 1.7.0 nor 1.7.1 is completely fine. 

Part of the issue is in https://docs.scipy.org/doc/ which contains all of the links. It links to:
- https://docs.scipy.org/doc/scipy-1.7.1/reference/
- https://docs.scipy.org/doc/scipy-1.7.0/reference/

A key difference is in the refguide link in the top navbar. 1.7.0 links to https://docs.scipy.org/doc/scipy-1.7.0/reference/index.html#, while 1.7.1 links to https://docs.scipy.org/doc/scipy-1.7.1/reference/reference/index.html which has the double `reference/reference` and does not render properly. I don't immediately understand why 1.7.1 seems to have regressed there.

Yes, the `objects.inv` is indeed in the wrong place there, that's caused by gh-14272. Or maybe it is in the place we want, but other projects are pointing at the old place. If we want https://docs.scipy.org/doc/scipy-1.7.1/ to be the front page, then that's the directory where `objects.inv` has to be - but numpy et al. are still pointing at https://docs.scipy.org/doc/scipy-1.7.1/reference/. My manual fix for 1.7.0 ensured we had `objects.inv` files in _both_ directories, but for 1.7.1 that's no longer the case.

So the question is: do we want to do this the "right" way, or the backwards compatible way?

**Comment by tupui** [comment]
Updated the name so we hopefully have less new issues about this. We could also pin the issue?

**Comment by heyitsguay** [comment]
This is still broken, I'm getting 404's on the search page and on getting started.

**Comment by tupui** [comment]
Yes @heyitsguay, the issue is still open... Please head to the developer website in the meantime https://scipy.github.io/devdocs/index.html

**Comment by rgommers** [comment]
Moving to 1.8.0, let's not mix it with the release that introduces M1 support. Best not to upload any new docs for `1.7.3` (just like we didn't for `1.7.2`).

**Comment by rgommers** [comment]
Now that we're close to `1.8.0`, I would say this is a release blocker. We need to sort this out properly - if we don't do it now, we never will.

**Comment by tupui** [comment]
Can we exactly what is done in the CI with the dev doc? Ideally I think it would be good to have this in the CI. We could progressively move parts of the release process there to get closer to be able to do continuous development (not that we would necessarily want to do it).

**Comment by rgommers** [comment]
> Can we do exactly what is done in the CI with the dev doc? 

The issue is broken references, what you are suggesting here is equivalent to "let's stay with the new layout and ignore all broken incoming links". Which is one option, but I'd like to check if we have better ones.

> Ideally I think it would be good to have this in the CI.

We can't. There are keys to a server involved, and unless we move away from that server (which is hard, we've been over that multiple times) there's no way we are going to deploy there straight from CI.

**Comment by tupui** [comment]
ok thanks for the clarifications 👍

**Comment by rgommers** [comment]
Hmm, with the new Sphinx team the size of the html docs has really exploded, making it a lot harder to migrate.

```
$ du -sh doc
7.5G    doc
$du -sh doc/numpy-1.17.0/
58M     doc/numpy-1.17.0/
$ du -sh doc/scipy-1.5.4
158M    doc/scipy-1.5.4
$ du -sh doc/scipy-1.6.3
162M    doc/scipy-1.6.3
$ du -sh doc/scipy-1.7.1
699M    doc/scipy-1.7.1
```
Although I suspect that's an issue we have fixed in the meantime (the sidebar change), because on a local build on `master` I get `162M` - the same as for `1.6.3`.

**Comment by rgommers** [comment]
I'm thinking we should add redirects manually. We have two `.htaccess` files, one on https://docs.scipy.org/ (that has a single blanket rule to forward to https://docs.scipy.org/doc/), and one on https://docs.scipy.org/doc/. That second one has only one rule:
```
/srv/docs_scipy_org/doc$ cat .htaccess 
RewriteEngine on
RewriteRule ^numpy(/.*) https://numpy.org/doc/stable$1 [R=permanent]
```
I think we need two redirects:
1. For `https://docs.scipy.org/doc/scipy/reference/objects.inv`, mapping to `https://docs.scipy.org/doc/scipy/objects.inv`.
2. For all non-existing URLs under `doc/source/reference/`, to redirect to the front page of the most recent version (`doc/scipy/index.html`)

And then we can deploy the `1.8.0` docs with only the structure we have on http://scipy.github.io/devdocs/.

Does that seem reasonable?

**Comment by rgommers** [comment]
I think this should be the new content of `doc/.htaccess`:
```
Options +SymLinksIfOwnerMatch 
RewriteEngine On 
RewriteRule ^numpy(/.*) https://numpy.org/doc/stable$1 [R=permanent]
# See https://github.com/scipy/scipy/issues/14267
Redirect 301 /scipy/reference/objects.inv /scipy/objects.inv
RewriteCond %{REQUEST_FILENAME} !-f 
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule . /index.html [L]
```
I basically never do this, so I'd appreciate if someone with some more expertise could review this.

**Comment by tylerjereddy** [comment]
Re: the size increase--perhaps worth noting that we both (or at least I did) copied various assets across a few different folders trying to manually patch broken links for the avalanche of issues that popped up on the issue tracker after switching to the new docs theme.

It still strikes me as challenging to reach a reasonable level of confidence that we don't break something, I'd never even heard of an `.htaccess` file before now. Worth a try I suppose.

**Comment by tylerjereddy** [comment]
@rgommers is `.htaccess` version controlled somewhere? Or you mean directly editing a file that is on the docs server, without changing anything else?

**Comment by rgommers** [comment]
> Or you mean directly editing a file that is on the docs server, without changing anything else?

Yep, this. Not in version control anywhere. Servers are fun ....

**Comment by tylerjereddy** [comment]
I'm backing up the changes in version control here: https://github.com/tylerjereddy/htaccess_backup 

Perpare for a potential cascade of intersphinx-related complaints if anything is off. But we likely can't fix the problem if we don't try a suggestion at some point.

**Comment by tylerjereddy** [comment]
`/srv/docs_scipy_org/doc/.htaccess` on the docs server has been adjusted to reflect the suggestion from Ralf above, as a first pass.

**Comment by adeak** [comment]
Is https://docs.scipy.org/doc/scipy/reference/ being a 404 the same issue?

And is https://docs.scipy.org/doc/numpy/reference/ saying "moved temporary" with a broken redirect and 404 target link https://docs.scipy.org/doc/numpy/reference/ related, or does that need reporting with NumPy?

**Comment by kolibril13** [comment]
Currently, it is difficult to reach the scipy docs from google:

https://user-images.githubusercontent.com/44469195/152681187-21be023a-3930-45e2-bfeb-417343881b19.mp4

**Comment by rossbar** [comment]
It looks like the intersphinx links to scipy from the numpy docs are now broken as well, see numpy/numpy#21004

**Comment by tylerjereddy** [comment]
Concrete suggestions for improving `/srv/docs_scipy_org/doc/.htaccess` are welcome. Otherwise, I'll probably just wait for some other core devs to make suggestions next week.

**Comment by mangecoeur** [comment]
> Is https://docs.scipy.org/doc/scipy/reference/ being a 404 the same issue?
> 
> And is https://docs.scipy.org/doc/numpy/reference/ saying "moved permanently" with a broken redirect seemingly to itself related, or does that need reporting with NumPy?
> 
> I found both links from https://docs.scipy.org/doc/.

Likewise seeing 404 on links from google. For me only the 'developer reference' links pointing to http://scipy.github.io/devdocs/dev/ are working.

**Comment by rgommers** [comment]
@tylerjereddy looks like the fire is out? You released 1.8.0 while I was offline for a few days, so I missed most of the fun. If there's something specific you want me to look at for this issue, please let me know.

**Comment by rgommers** [comment]
Ahh actually no, something went wrong during upload @tylerjereddy. On the doc server, the content that should be in `/srv/docs_scipy_org/doc/scipy/` is under `/srv/docs_scipy_org/doc/html-scipyorg/`. That was never intended.

On the `numpy-tutorials` repo it looks like the intersphinx link was changed to `'scipy': ('https://docs.scipy.org/doc/scipy', None),` which is what we want. It just seems like we're uploading to `html-scipyorg/` (not what we want) and then running the redirects in reverse.

**Comment by tylerjereddy** [comment]
> Ahh actually no, something went wrong during upload

Really? I still have the local build artifacts from `make dist` and the structure seems to match, and I'm pretty sure the `make upload` type command exited with `0`.

![image](https://user-images.githubusercontent.com/7903078/153334723-7a972f40-44b6-45f7-ae13-3f81b78f46f5.png)

**Comment by tupui** [comment]
I am re-pinning the issue, we keep getting bug reports otherwise.

**Comment by bskinn** [comment]
As best as I'm able to determine, the `intersphinx_mapping` entry I've placed at [this Gist](https://gist.github.com/bskinn/0e164963428d4b51017cebdb6cda5209) *does work* for the current hosting situation for 1.8.0.

```
intersphinx_mapping = {"scipy": ("https://docs.scipy.org/doc/scipy-1.8.0/html-scipyorg/", None)}
```

It [passes](https://github.com/bskinn/intersphinx-gist/runs/5159987133?check_suite_focus=true) the tests I've written for the mappings at the Gist (see the "Report Log" workflow step); and, a stub Sphinx project I spun up just now _is_ able to pull the `objects.inv` and correctly insert a cross-ref to the `scipy.sparse.linalg.SuperLU` class.

I'll be keeping the mapping entries at that Gist as up to date as I can, as the work proceeds here.

**Comment by rgommers** [comment]
This broke again after the 1.8.1 release. It should finally be fixed for good by:
- gh-16229
- gh-16221 (same PR for 1.8.x branch, already deployed)
- https://github.com/scipy/docs.scipy.org/pull/64 (already deployed)

If anyone still sees something that's not in the correct place, please comment.

**Comment by rgommers** [comment]
It seems we've finally gotten this fixed in the release scripts now. The stable intersphinx link that I'd expect project to use is:
```
scipy': ('https://docs.scipy.org/doc/scipy', None),
```
This has always worked, with short hiccups after the 1.8.0 and 1.8.1 releases. It shouldn't be happening after the next release anymore.
